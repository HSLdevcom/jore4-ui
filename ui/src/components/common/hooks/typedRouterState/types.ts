import { Dispatch, SetStateAction } from 'react';

export type UrlStateSerializer<T> = (value: T) => string;
export type UrlStateDeserializer<T> = (value: string) => T;

/**
 * Mapping from a state object to a collection of transformer functions
 * needed to stringify the individual fields of the state object.
 */
export type UrlStateSerializers<StateT extends object> = {
  readonly [key in keyof StateT]: UrlStateSerializer<StateT[key]>;
};

/**
 * Mapping from a state object to a collection of transformer functions
 * needed to retrieve the actual value of the state objects field,
 * from its string representation.
 */
export type UrlStateDeserializers<StateT extends object> = {
  readonly [key in keyof StateT]: UrlStateDeserializer<StateT[key]>;
};

/**
 * Helper type used when reassembling the state from the URl params field by field.
 */
export type PartialState = Record<string, unknown>;

export type SimpleType =
  | string
  | number
  | boolean
  | ReadonlyArray<SimpleType>
  | { readonly [key: string]: SimpleType };

export type SimpleRecord = Readonly<Record<string, SimpleType>>;

export type TypedRouterState<
  SearchStateT extends object,
  HistoryStateT extends SimpleRecord,
> = {
  readonly search: SearchStateT;
  readonly history: HistoryStateT;
};

export type TypedRouterStateStateSetters<
  SearchStateT extends object,
  HistoryStateT extends SimpleRecord,
> = {
  readonly setState: Dispatch<
    SetStateAction<TypedRouterState<SearchStateT, HistoryStateT>>
  >;
  readonly setSearchState: Dispatch<SetStateAction<SearchStateT>>;
  readonly setHistoryState: Dispatch<SetStateAction<HistoryStateT>>;
};

export type TypedRouterStateHookResponse<
  SearchStateT extends object,
  HistoryStateT extends SimpleRecord,
> = {
  readonly state: TypedRouterState<SearchStateT, HistoryStateT>;
} & TypedRouterStateStateSetters<SearchStateT, HistoryStateT>;
