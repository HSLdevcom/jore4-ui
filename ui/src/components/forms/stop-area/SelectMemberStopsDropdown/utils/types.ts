import { GetStopPointsByQuayIdQuery } from '../../../../../generated/graphql';

export type MoveQuayParams = {
  toStopPlaceId: string;
  quayIds: string[];
  moveQuayFromDate: string;
  fromVersionComment: string;
  toVersionComment: string;
};

export type QuayInfo = {
  id: string;
  publicCode: string;
};

export type StopPointInfo =
  GetStopPointsByQuayIdQuery['service_pattern_scheduled_stop_point'][0];

export type MoveStopPlace = {
  __typename?: string;
  id?: string | null;
  version?: string | null;
  versionComment?: string | null;
  quays?: Array<{
    __typename?: string;
    id?: string | null;
    publicCode?: string | null;
    keyValues?: Array<{
      __typename?: string;
      key?: string | null;
      values?: Array<string | null> | null;
    } | null> | null;
  } | null> | null;
};
