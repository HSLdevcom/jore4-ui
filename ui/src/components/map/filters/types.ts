export type FilterItem = {
  readonly id: string;
  readonly isActive: boolean;
  readonly label: string;
  readonly toggleFunction: (isActive: boolean) => void;
  readonly disabled: boolean;
};

export type ClassNameProps = {
  readonly className?: string;
};
