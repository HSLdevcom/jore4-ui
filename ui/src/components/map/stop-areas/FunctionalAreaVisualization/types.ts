export type PixelCoordinates = [number, number]; // X, Y

// Geojson coordinates [longitude, latitude]
// GeoJSON package only provides an alias for anysized number[] array.
export type Coordinates = [number, number];

export type CenterRadius = {
  readonly center: Coordinates;
  readonly radius: number;
};

export type Circles = {
  readonly stops: ReadonlyArray<CenterRadius>;
  readonly area: CenterRadius;
};

export interface PixelCoordinateConverter {
  coordinates(coordinates: Coordinates): PixelCoordinates;
  radius(radius: number): number;
}
