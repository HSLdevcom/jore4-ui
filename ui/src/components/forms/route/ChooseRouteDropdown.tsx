import { DateTime } from 'luxon';
import { useState } from 'react';
import { useTranslation } from 'react-i18next';
import { RouteAllFieldsFragment } from '../../../generated/graphql';
import { useChooseRouteDropdown } from '../../../hooks';
import { MAX_DATE, MIN_DATE } from '../../../time';
import { Priority } from '../../../types/enums';
import { ComboboxInputProps, SearchableDropdown } from '../../../uiComponents';
import { DateRange } from '../common/DateRange';

interface Props extends ComboboxInputProps {
  testId?: string;
  date: DateTime;
  priorities: Priority[];
}

const mapToOptionContent = (item: RouteAllFieldsFragment) => (
  <div className="flex flex-col">
    <div>
      <span className="font-bold">{item.label}</span>
      {` | ${item.name_i18n?.fi_FI}`}
    </div>
    <div className="text-sm">
      <DateRange
        startDate={item.validity_start || MIN_DATE}
        endDate={item.validity_end || MAX_DATE}
      />
    </div>
  </div>
);

const mapToOption = (item: RouteAllFieldsFragment) => ({
  key: item.route_id,
  value: item.route_id,
  render: () => mapToOptionContent(item),
});

export const ChooseRouteDropdown = ({
  testId,
  value,
  onChange,
  onBlur,
  date,
  priorities,
}: Props): JSX.Element => {
  const { t } = useTranslation();

  const [query, setQuery] = useState('');

  const { routes, selectedRoute } = useChooseRouteDropdown({
    query,
    observationDate: date,
    priorities,
    routeId: value,
  });

  const options = routes?.map(mapToOption) || [];

  const mapToButtonContent = (displayedRoute?: RouteAllFieldsFragment) => {
    // If no route is selected, show "Choose route"
    return (
      <div className="w-full">
        {displayedRoute
          ? `${displayedRoute.label} (${displayedRoute.name_i18n?.fi_FI})`
          : t('routes.chooseRoute')}
      </div>
    );
  };

  return (
    <SearchableDropdown
      id="choose-route-combobox"
      testId={testId}
      mapToButtonContent={mapToButtonContent}
      options={options}
      value={value}
      onChange={onChange}
      onBlur={onBlur}
      onQueryChange={setQuery}
      selectedItem={selectedRoute}
    />
  );
};
