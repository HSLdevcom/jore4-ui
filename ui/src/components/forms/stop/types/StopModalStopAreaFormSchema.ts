import { z } from 'zod';

const optionalString = z.string().nullable();

export const stopModalStopAreaFormSchema = z.object({
  netextId: z.string(),
  privateCode: z.string(),

  validityStart: z.string(),
  validityEnd: optionalString,

  nameFin: optionalString,
  nameSwe: optionalString,

  longNameFin: optionalString,
  longNameSwe: optionalString,

  abbreviationFin: optionalString,
  abbreviationSwe: optionalString,
});

export type StopModalStopAreaFormSchema = z.infer<
  typeof stopModalStopAreaFormSchema
>;
