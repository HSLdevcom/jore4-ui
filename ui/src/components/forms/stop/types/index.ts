export * from './StopFormSchema';
export * from './StopModalStopAreaFormSchema';
