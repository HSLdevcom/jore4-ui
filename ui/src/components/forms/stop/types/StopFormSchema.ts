import { z } from 'zod';
import {
  REQUIRED_FIELD_ERROR_MESSAGE,
  schema as changeValidityFormSchema,
  refineValidityPeriodSchema,
  requiredNumber,
  requiredString,
} from '../../common';
import { stopModalStopAreaFormSchema } from './StopModalStopAreaFormSchema';

export const MISSING_ID = 'MISSING_ID';

const publicCodeSchema = z.object({
  value: requiredString,
  municipality: z.string().nullable(),
  expectedPrefix: z.string().nullable(),
});

export type StopPublicCodeState = z.infer<typeof publicCodeSchema>;

const keyValueSchema = z.object({
  key: requiredString,
  values: z.array(z.string().nullable()),
});

export const stopFormSchema = z
  .object({
    stopId: z.string().uuid().optional(), // for stops that are edited
    quayId: z.string().optional(), // for stops that are edited
    stopArea: stopModalStopAreaFormSchema.nullable(),
    publicCode: publicCodeSchema.required(),
    latitude: requiredNumber.min(-180).max(180),
    longitude: requiredNumber.min(-180).max(180),
    locationFin: z.string().nullable().optional(),
    locationSwe: z.string().nullable().optional(),
    timingPlaceId: z.string().uuid().nullable().optional(),
    versionName: z.string().nullable().optional(),
    versionDescription: z.string().nullable().optional(),
    keyValues: z.array(keyValueSchema).optional(),
  })
  .merge(changeValidityFormSchema)
  .superRefine((stop, ctx) => {
    // We should either have both a StopPoint and a Quay ID, or neither.
    if (!!stop.stopId !== !!stop.quayId) {
      ctx.addIssue({
        code: 'custom',
        message: MISSING_ID,
        path: ['stopId'],
      });
    }

    // If we don't have a StopArea selected
    if (!stop.stopArea?.netextId) {
      ctx.addIssue({
        code: 'custom',
        message: REQUIRED_FIELD_ERROR_MESSAGE,
        path: ['stopArea'],
      });
    }
  })
  .superRefine(refineValidityPeriodSchema);

export type StopFormState = z.infer<typeof stopFormSchema>;
