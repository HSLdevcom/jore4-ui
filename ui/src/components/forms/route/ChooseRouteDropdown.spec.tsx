import { MockedProvider, MockedResponse } from '@apollo/client/testing';
import { buildRoute } from '@hsl/jore4-test-db-manager';
import { act, fireEvent, screen } from '@testing-library/react';
import { DateTime } from 'luxon';
import {
  GetRouteDetailsByLabelWildcardDocument,
  GetRouteDetailsByLabelWildcardQuery,
  RouteDirectionEnum,
} from '../../../generated/graphql';
import { Priority } from '../../../types/enums';
import {
  fireFullMouseClickSequence,
  render,
  sleep,
} from '../../../utils/test-utils';
import { ChooseRouteDropdown } from './ChooseRouteDropdown';

describe(`<${ChooseRouteDropdown.name} />`, () => {
  // eslint-disable-next-line @typescript-eslint/no-empty-function
  const onChange = () => {};
  // eslint-disable-next-line @typescript-eslint/no-empty-function
  const onBlur = () => {};
  const value = '';
  const testId = 'ChooseRouteDropdown1';
  const buttonTestId = `${testId}::button`;
  const inputTestId = `${testId}::input`;

  const mocks: MockedResponse<GetRouteDetailsByLabelWildcardQuery>[] = [
    {
      request: {
        query: GetRouteDetailsByLabelWildcardDocument,
        variables: {
          labelPattern: '%',
          date: '2022-03-23T00:00:00.000+02:00',
          priorities: [10],
        },
      },
      result: {
        data: {
          route_route: [
            {
              ...buildRoute({ label: '123' }),
              __typename: 'route_route',
              route_id: 'route1',
              validity_start: DateTime.fromISO('2022-03-20T22:00:00+00:00'),
              validity_end: DateTime.fromISO('2022-03-29T20:59:59.999+00:00'),
              priority: Priority.Standard,
              direction: RouteDirectionEnum.Outbound,
              route_shape: null,
              on_line_id: 'line_id',
            },
            {
              ...buildRoute({ label: '456' }),
              __typename: 'route_route',
              route_id: 'route2',
              validity_start: DateTime.fromISO('2021-01-01T22:00:00+00:00'),
              validity_end: DateTime.fromISO('2022-12-31T21:59:59.999+00:00'),
              priority: Priority.Standard,
              direction: RouteDirectionEnum.Outbound,
              route_shape: null,
              on_line_id: 'line_id',
            },
          ],
        },
      },
    },
    {
      request: {
        query: GetRouteDetailsByLabelWildcardDocument,
        variables: {
          labelPattern: '1%',
          date: '2022-03-23T00:00:00.000+02:00',
          priorities: [10],
        },
      },
      result: {
        data: {
          route_route: [
            {
              ...buildRoute({ label: '123' }),
              label: '123',
              __typename: 'route_route',
              route_id: 'route1',
              validity_start: DateTime.fromISO('2022-03-20T22:00:00+00:00'),
              validity_end: DateTime.fromISO('2022-03-29T20:59:59.999+00:00'),
              priority: Priority.Standard,
              direction: RouteDirectionEnum.Outbound,
              route_shape: null,
              on_line_id: 'line_id',
            },
          ],
        },
      },
    },
  ];

  test('Opens dropdown when clicked and shows all routes', async () => {
    const { asFragment } = render(
      <MockedProvider mocks={mocks} addTypename={false}>
        <ChooseRouteDropdown
          testId={testId}
          value={value}
          onChange={onChange}
          onBlur={onBlur}
          date={DateTime.fromISO('2022-03-23T00:00:00.000+02:00')}
          priorities={[Priority.Standard]}
        />
      </MockedProvider>,
    );

    // dropdown is collapsed, value is not yet bound to dropdown, 'Choose route' text shows
    expect(screen.getByTestId(buttonTestId)).toHaveTextContent(
      'Valitse reitti',
    );
    expect(screen.queryAllByRole('option')).toHaveLength(0);
    expect(asFragment()).toMatchSnapshot();

    // wait for the graphql call to execute
    await act(() => sleep(0));

    // dropdown is collapsed, data is loaded and 'Choose route' text still shows in button
    expect(screen.getByTestId(buttonTestId)).toHaveTextContent(
      'Valitse reitti',
    );
    expect(screen.queryAllByRole('option')).toHaveLength(0);
    expect(asFragment()).toMatchSnapshot();

    // click dropdown to open it:
    const openDropdownButton = screen.getByTestId(buttonTestId);
    fireFullMouseClickSequence(openDropdownButton);

    // dropdown is open, both lines show
    const items = screen.queryAllByRole('option');

    expect(items[0].querySelectorAll('div > div')[0].textContent).toBe(
      `123 | route 123`,
    );
    expect(items[0].querySelectorAll('div > div')[1].textContent).toBe(
      `21.3.2022 - 29.3.2022`,
    );
    expect(items[1].querySelectorAll('div > div')[0].textContent).toBe(
      `456 | route 456`,
    );
    expect(items[1].querySelectorAll('div > div')[1].textContent).toBe(
      `2.1.2021 - 31.12.2022`,
    );
    expect(items.length).toBe(2);

    expect(asFragment()).toMatchSnapshot();
  });

  test('Filters shown routes when query is set', async () => {
    const { asFragment } = render(
      <MockedProvider mocks={mocks} addTypename={false}>
        <ChooseRouteDropdown
          testId={testId}
          value={value}
          onChange={onChange}
          onBlur={onBlur}
          date={DateTime.fromISO('2022-03-23T00:00:00.000+02:00')}
          priorities={[Priority.Standard]}
        />
      </MockedProvider>,
    );

    // dropdown is collapsed, value is not yet bound to dropdown, 'Choose route' text shows
    expect(screen.getByTestId(buttonTestId)).toHaveTextContent(
      'Valitse reitti',
    );
    expect(screen.queryAllByRole('option')).toHaveLength(0);
    expect(asFragment()).toMatchSnapshot();

    // wait for the graphql call to execute
    await act(() => sleep(0));

    // dropdown is collapsed, data is loaded and 'Choose route' text still shows in button
    expect(screen.getByTestId(buttonTestId)).toHaveTextContent(
      'Valitse reitti',
    );
    expect(screen.queryAllByRole('option')).toHaveLength(0);
    expect(asFragment()).toMatchSnapshot();

    // write a query to the input textbox
    const input = screen.getByTestId(inputTestId);
    fireEvent.change(input, { target: { value: '1' } });

    // To make sure that the query result is ready, we need to use real timers again
    await act(async () => sleep(0));
    // dropdown is open, first line shows
    const items = screen.queryAllByRole('option');

    expect(items[0].querySelectorAll('div > div')[0].textContent).toBe(
      `123 | route 123`,
    );
    expect(items[0].querySelectorAll('div > div')[1].textContent).toBe(
      `21.3.2022 - 29.3.2022`,
    );
    expect(items.length).toBe(1);

    expect(asFragment()).toMatchSnapshot();
  });
});
