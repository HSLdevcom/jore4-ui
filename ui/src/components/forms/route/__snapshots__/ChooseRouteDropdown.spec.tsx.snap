// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ChooseRouteDropdown /> Filters shown routes when query is set 1`] = `
<DocumentFragment>
  <div
    class="relative w-full"
    data-headlessui-state=""
    data-testid="ChooseRouteDropdown1"
    id="choose-route-combobox"
  >
    <input
      aria-autocomplete="list"
      aria-expanded="false"
      class="relative h-full w-full rounded-md border border-grey bg-white py-3 px-2"
      data-headlessui-state=""
      data-testid="ChooseRouteDropdown1::input"
      id="headlessui-combobox-input-:r5:"
      role="combobox"
      type="text"
      value=""
    />
    <button
      aria-expanded="false"
      aria-haspopup="listbox"
      class="absolute inset-y-0 right-0 flex h-full w-full items-center justify-end px-3 text-left focus:outline-none"
      data-headlessui-state=""
      data-testid="ChooseRouteDropdown1::button"
      id="headlessui-combobox-button-:r6:"
      tabindex="-1"
      type="button"
    >
      <div
        class="w-full"
      >
        Valitse reitti
      </div>
      <svg
        class="mr-1 ml-auto text-2xl text-tweaked-brand"
        fill="currentColor"
        height="1em"
        stroke="currentColor"
        stroke-width="0"
        viewBox="0 0 24 24"
        width="1em"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M0 0h24v24H0z"
          fill="none"
        />
        <path
          d="M15.5 14h-.79l-.28-.27A6.471 6.471 0 0016 9.5 6.5 6.5 0 109.5 16c1.61 0 3.09-.59 4.23-1.57l.27.28v.79l5 4.99L20.49 19l-4.99-5zm-6 0C7.01 14 5 11.99 5 9.5S7.01 5 9.5 5 14 7.01 14 9.5 11.99 14 9.5 14z"
        />
      </svg>
      <i
        class="icon-arrow text-tweaked-brand transition duration-150 ease-in-out rotate-0"
        style="font-size: 10px;"
      />
    </button>
  </div>
</DocumentFragment>
`;

exports[`<ChooseRouteDropdown /> Filters shown routes when query is set 2`] = `
<DocumentFragment>
  <div
    class="relative w-full"
    data-headlessui-state=""
    data-testid="ChooseRouteDropdown1"
    id="choose-route-combobox"
  >
    <input
      aria-autocomplete="list"
      aria-expanded="false"
      class="relative h-full w-full rounded-md border border-grey bg-white py-3 px-2"
      data-headlessui-state=""
      data-testid="ChooseRouteDropdown1::input"
      id="headlessui-combobox-input-:r5:"
      role="combobox"
      type="text"
      value=""
    />
    <button
      aria-expanded="false"
      aria-haspopup="listbox"
      class="absolute inset-y-0 right-0 flex h-full w-full items-center justify-end px-3 text-left focus:outline-none"
      data-headlessui-state=""
      data-testid="ChooseRouteDropdown1::button"
      id="headlessui-combobox-button-:r6:"
      tabindex="-1"
      type="button"
    >
      <div
        class="w-full"
      >
        Valitse reitti
      </div>
      <svg
        class="mr-1 ml-auto text-2xl text-tweaked-brand"
        fill="currentColor"
        height="1em"
        stroke="currentColor"
        stroke-width="0"
        viewBox="0 0 24 24"
        width="1em"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M0 0h24v24H0z"
          fill="none"
        />
        <path
          d="M15.5 14h-.79l-.28-.27A6.471 6.471 0 0016 9.5 6.5 6.5 0 109.5 16c1.61 0 3.09-.59 4.23-1.57l.27.28v.79l5 4.99L20.49 19l-4.99-5zm-6 0C7.01 14 5 11.99 5 9.5S7.01 5 9.5 5 14 7.01 14 9.5 11.99 14 9.5 14z"
        />
      </svg>
      <i
        class="icon-arrow text-tweaked-brand transition duration-150 ease-in-out rotate-0"
        style="font-size: 10px;"
      />
    </button>
  </div>
</DocumentFragment>
`;

exports[`<ChooseRouteDropdown /> Filters shown routes when query is set 3`] = `
<DocumentFragment>
  <div
    class="relative w-full"
    data-headlessui-state="open"
    data-testid="ChooseRouteDropdown1"
    id="choose-route-combobox"
  >
    <input
      aria-activedescendant="headlessui-combobox-option-:r8:"
      aria-autocomplete="list"
      aria-controls="headlessui-combobox-options-:r7:"
      aria-expanded="true"
      class="relative h-full w-full rounded-md border border-grey bg-white py-3 px-2"
      data-headlessui-state="open"
      data-testid="ChooseRouteDropdown1::input"
      id="headlessui-combobox-input-:r5:"
      role="combobox"
      type="text"
      value=""
    />
    <button
      aria-controls="headlessui-combobox-options-:r7:"
      aria-expanded="true"
      aria-haspopup="listbox"
      class="absolute inset-y-0 right-0 flex h-full w-full items-center justify-end px-3 text-left focus:outline-none"
      data-headlessui-state="open"
      data-testid="ChooseRouteDropdown1::button"
      id="headlessui-combobox-button-:r6:"
      tabindex="-1"
      type="button"
    >
      <svg
        class="mr-1 ml-auto text-2xl text-tweaked-brand"
        fill="currentColor"
        height="1em"
        stroke="currentColor"
        stroke-width="0"
        viewBox="0 0 24 24"
        width="1em"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M0 0h24v24H0z"
          fill="none"
        />
        <path
          d="M15.5 14h-.79l-.28-.27A6.471 6.471 0 0016 9.5 6.5 6.5 0 109.5 16c1.61 0 3.09-.59 4.23-1.57l.27.28v.79l5 4.99L20.49 19l-4.99-5zm-6 0C7.01 14 5 11.99 5 9.5S7.01 5 9.5 5 14 7.01 14 9.5 11.99 14 9.5 14z"
        />
      </svg>
      <i
        class="icon-arrow text-tweaked-brand transition duration-150 ease-in-out -rotate-180"
        style="font-size: 10px;"
      />
    </button>
    <ul
      aria-labelledby="headlessui-combobox-button-:r6:"
      class="absolute left-0 z-10 w-full rounded-b-md border border-black border-opacity-20 bg-white shadow-md focus:outline-none transition ease-out duration-100 opacity-0 scale-95"
      data-headlessui-state="open"
      id="headlessui-combobox-options-:r7:"
      role="listbox"
    >
      <li
        aria-selected="false"
        data-headlessui-state="active"
        id="headlessui-combobox-option-:r8:"
        role="option"
        tabindex="-1"
      >
        <span
          class="bg-background flex group text-left pr-2 py-2 focus:outline-none items-center"
        >
          <svg
            class="mx-1 text-2xl invisible"
            fill="currentColor"
            height="1em"
            stroke="currentColor"
            stroke-width="0"
            viewBox="0 0 24 24"
            width="1em"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M0 0h24v24H0z"
              fill="none"
            />
            <path
              d="M9 16.17L4.83 12l-1.42 1.41L9 19 21 7l-1.41-1.41z"
            />
          </svg>
          <div
            class="flex flex-col"
          >
            <div>
              <span
                class="font-bold"
              >
                123
              </span>
               | route 123
            </div>
            <div
              class="text-sm"
            >
              21.3.2022 - 29.3.2022
            </div>
          </div>
        </span>
      </li>
    </ul>
  </div>
</DocumentFragment>
`;

exports[`<ChooseRouteDropdown /> Opens dropdown when clicked and shows all routes 1`] = `
<DocumentFragment>
  <div
    class="relative w-full"
    data-headlessui-state=""
    data-testid="ChooseRouteDropdown1"
    id="choose-route-combobox"
  >
    <input
      aria-autocomplete="list"
      aria-expanded="false"
      class="relative h-full w-full rounded-md border border-grey bg-white py-3 px-2"
      data-headlessui-state=""
      data-testid="ChooseRouteDropdown1::input"
      id="headlessui-combobox-input-:r0:"
      role="combobox"
      type="text"
      value=""
    />
    <button
      aria-expanded="false"
      aria-haspopup="listbox"
      class="absolute inset-y-0 right-0 flex h-full w-full items-center justify-end px-3 text-left focus:outline-none"
      data-headlessui-state=""
      data-testid="ChooseRouteDropdown1::button"
      id="headlessui-combobox-button-:r1:"
      tabindex="-1"
      type="button"
    >
      <div
        class="w-full"
      >
        Valitse reitti
      </div>
      <svg
        class="mr-1 ml-auto text-2xl text-tweaked-brand"
        fill="currentColor"
        height="1em"
        stroke="currentColor"
        stroke-width="0"
        viewBox="0 0 24 24"
        width="1em"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M0 0h24v24H0z"
          fill="none"
        />
        <path
          d="M15.5 14h-.79l-.28-.27A6.471 6.471 0 0016 9.5 6.5 6.5 0 109.5 16c1.61 0 3.09-.59 4.23-1.57l.27.28v.79l5 4.99L20.49 19l-4.99-5zm-6 0C7.01 14 5 11.99 5 9.5S7.01 5 9.5 5 14 7.01 14 9.5 11.99 14 9.5 14z"
        />
      </svg>
      <i
        class="icon-arrow text-tweaked-brand transition duration-150 ease-in-out rotate-0"
        style="font-size: 10px;"
      />
    </button>
  </div>
</DocumentFragment>
`;

exports[`<ChooseRouteDropdown /> Opens dropdown when clicked and shows all routes 2`] = `
<DocumentFragment>
  <div
    class="relative w-full"
    data-headlessui-state=""
    data-testid="ChooseRouteDropdown1"
    id="choose-route-combobox"
  >
    <input
      aria-autocomplete="list"
      aria-expanded="false"
      class="relative h-full w-full rounded-md border border-grey bg-white py-3 px-2"
      data-headlessui-state=""
      data-testid="ChooseRouteDropdown1::input"
      id="headlessui-combobox-input-:r0:"
      role="combobox"
      type="text"
      value=""
    />
    <button
      aria-expanded="false"
      aria-haspopup="listbox"
      class="absolute inset-y-0 right-0 flex h-full w-full items-center justify-end px-3 text-left focus:outline-none"
      data-headlessui-state=""
      data-testid="ChooseRouteDropdown1::button"
      id="headlessui-combobox-button-:r1:"
      tabindex="-1"
      type="button"
    >
      <div
        class="w-full"
      >
        Valitse reitti
      </div>
      <svg
        class="mr-1 ml-auto text-2xl text-tweaked-brand"
        fill="currentColor"
        height="1em"
        stroke="currentColor"
        stroke-width="0"
        viewBox="0 0 24 24"
        width="1em"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M0 0h24v24H0z"
          fill="none"
        />
        <path
          d="M15.5 14h-.79l-.28-.27A6.471 6.471 0 0016 9.5 6.5 6.5 0 109.5 16c1.61 0 3.09-.59 4.23-1.57l.27.28v.79l5 4.99L20.49 19l-4.99-5zm-6 0C7.01 14 5 11.99 5 9.5S7.01 5 9.5 5 14 7.01 14 9.5 11.99 14 9.5 14z"
        />
      </svg>
      <i
        class="icon-arrow text-tweaked-brand transition duration-150 ease-in-out rotate-0"
        style="font-size: 10px;"
      />
    </button>
  </div>
</DocumentFragment>
`;

exports[`<ChooseRouteDropdown /> Opens dropdown when clicked and shows all routes 3`] = `
<DocumentFragment>
  <div
    class="relative w-full"
    data-headlessui-state="open"
    data-testid="ChooseRouteDropdown1"
    id="choose-route-combobox"
  >
    <input
      aria-activedescendant="headlessui-combobox-option-:r3:"
      aria-autocomplete="list"
      aria-controls="headlessui-combobox-options-:r2:"
      aria-expanded="true"
      class="relative h-full w-full rounded-md border border-grey bg-white py-3 px-2"
      data-headlessui-state="open"
      data-testid="ChooseRouteDropdown1::input"
      id="headlessui-combobox-input-:r0:"
      role="combobox"
      type="text"
      value=""
    />
    <button
      aria-controls="headlessui-combobox-options-:r2:"
      aria-expanded="true"
      aria-haspopup="listbox"
      class="absolute inset-y-0 right-0 flex h-full w-full items-center justify-end px-3 text-left focus:outline-none"
      data-headlessui-state="open"
      data-testid="ChooseRouteDropdown1::button"
      id="headlessui-combobox-button-:r1:"
      tabindex="-1"
      type="button"
    >
      <div
        class="w-full"
      >
        Valitse reitti
      </div>
      <svg
        class="mr-1 ml-auto text-2xl text-tweaked-brand"
        fill="currentColor"
        height="1em"
        stroke="currentColor"
        stroke-width="0"
        viewBox="0 0 24 24"
        width="1em"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M0 0h24v24H0z"
          fill="none"
        />
        <path
          d="M15.5 14h-.79l-.28-.27A6.471 6.471 0 0016 9.5 6.5 6.5 0 109.5 16c1.61 0 3.09-.59 4.23-1.57l.27.28v.79l5 4.99L20.49 19l-4.99-5zm-6 0C7.01 14 5 11.99 5 9.5S7.01 5 9.5 5 14 7.01 14 9.5 11.99 14 9.5 14z"
        />
      </svg>
      <i
        class="icon-arrow text-tweaked-brand transition duration-150 ease-in-out -rotate-180"
        style="font-size: 10px;"
      />
    </button>
    <ul
      aria-labelledby="headlessui-combobox-button-:r1:"
      class="absolute left-0 z-10 w-full rounded-b-md border border-black border-opacity-20 bg-white shadow-md focus:outline-none transition ease-out duration-100 opacity-0 scale-95"
      data-headlessui-state="open"
      id="headlessui-combobox-options-:r2:"
      role="listbox"
    >
      <li
        aria-selected="false"
        data-headlessui-state="active"
        id="headlessui-combobox-option-:r3:"
        role="option"
        tabindex="-1"
      >
        <span
          class="bg-background flex group text-left pr-2 py-2 focus:outline-none items-center"
        >
          <svg
            class="mx-1 text-2xl invisible"
            fill="currentColor"
            height="1em"
            stroke="currentColor"
            stroke-width="0"
            viewBox="0 0 24 24"
            width="1em"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M0 0h24v24H0z"
              fill="none"
            />
            <path
              d="M9 16.17L4.83 12l-1.42 1.41L9 19 21 7l-1.41-1.41z"
            />
          </svg>
          <div
            class="flex flex-col"
          >
            <div>
              <span
                class="font-bold"
              >
                123
              </span>
               | route 123
            </div>
            <div
              class="text-sm"
            >
              21.3.2022 - 29.3.2022
            </div>
          </div>
        </span>
      </li>
      <li
        aria-selected="false"
        data-headlessui-state=""
        id="headlessui-combobox-option-:r4:"
        role="option"
        tabindex="-1"
      >
        <span
          class=" flex group text-left pr-2 py-2 focus:outline-none items-center"
        >
          <svg
            class="mx-1 text-2xl invisible"
            fill="currentColor"
            height="1em"
            stroke="currentColor"
            stroke-width="0"
            viewBox="0 0 24 24"
            width="1em"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M0 0h24v24H0z"
              fill="none"
            />
            <path
              d="M9 16.17L4.83 12l-1.42 1.41L9 19 21 7l-1.41-1.41z"
            />
          </svg>
          <div
            class="flex flex-col"
          >
            <div>
              <span
                class="font-bold"
              >
                456
              </span>
               | route 456
            </div>
            <div
              class="text-sm"
            >
              2.1.2021 - 31.12.2022
            </div>
          </div>
        </span>
      </li>
    </ul>
  </div>
</DocumentFragment>
`;
